src/proj.rs

```diff
@@ -699,20 +732,43 @@ mod test {
     }
     // This test is disabled by default as it requires network access
     // #[test]
-    // fn test_network() {
+    // fn test_network_default_context() {
     //     let from = "EPSG:4326";
     //     let to = "EPSG:4326+3855";
     //     // off by default
-    //     assert_eq!(network_enabled(), false);
+    //     assert_eq!(ProjContext::DefaultContext.network_enabled(), false);
     //     // switch it on and disable cache for subsequent calls
-    //     grid_cache_set_enable(false);
-    //     enable_network(true).unwrap();
+    //     ProjContext::DefaultContext.grid_cache_set_enable(false);
```

Comment:
This API would now be more verbose - you have to specify `ProjContext::DefaultContext`.

On the plus side the "you must set network hooks before you instantate your `Proj` need not apply if users just state their Proj instance directly, as shown in the subsequent example.

OTOH if it's, more common for users to create lots of Proj instances throughout their code, maybe mutating the default context is actually what they'd want and we should steer them toward that as you had it in your original PR. 🤷

Reply:
Unless I'm missing something, this isolates an individual context from the global context when querying network-download state (etc), but doesn't allow the creation of an individual network-enabled transformation or projection instance, because that requires an individual context: you have to enable network functionality using `proj_context_set_enable_network` before you call either `proj_create` or `proj_create_crs_to_crs` – doing it afterwards is a no-op.

Reply:
The more I think about it, the more I become convinced that I've needlessly complicated this by not mirroring the PROJ API. I think a well-documented two step process (create a context, and modify it to enable downloads or new search paths if need be, then create the transformation / projection, consuming the context) is a clearer approach.

Reply:
>  you have to enable network functionality using proj_context_set_enable_network before you call either proj_create or proj_create_crs_to_crs – doing it afterwards is a no-op.

I confess that I don't really understand what the network features are used for (which is why I was looking into the network test examples yesterday). Going off my intuition and a cursory look at the proj.cpp code such as this:

https://github.com/OSGeo/PROJ/blob/master/src/grids.cpp#L3125

Which ultimately *can* open a file via the network file manager... it seemed like it was dynamically consulting the context referenced by the proj instance at the time you make the potential network call - not some state that was captured during proj_create. But maybe that's not true everywhere. 

If you provided an example of some rust code that uses the networking faculties, I could look into it more.

> The more I think about it, the more I become convinced that I've needlessly complicated this by not mirroring the PROJ API. I think a well-documented two step process (create a context, and modify it to enable downloads or new search paths if need be, then create the transformation / projection, consuming the context) is a clearer approach.

That would be safer and more explicit, though I do like the simplicity of your current API given that many (most?) folks won't need to touch their context (right?). I'm still hopeful that we could reconcile the two such that the complexity of "a context" is only disclosed progressively if folks need it.

Reply:
And I'd like to reiterate that you should follow your heart here rather than trying to please me - you've been shepherding this project for a while and have a better grasp on what its users want.

Reply:
In the original PR, I've now achieved what I think is a reasonable compromise:

1. First, these changes do not break the existing API: `Proj::new()` and `Proj::new_known_crs()` are untouched, and behave as they always have;
3. For users who need grids from the network, or to alter search paths or grid urls, there's a `TransformBuilder`;
4. This can be used to customise the properties of the current `PROJ` context – never the global properties;
5. `TransformBuilder` has two methods for generating a `Proj` instance: `new_transform`, and `transform_known_crs`. They have the same signatures as `Proj::new` and `Proj::new_known_crs` respectively. Calling them consumes the `TransformBuilder`, returning a `Proj` instance;
6. Both `Proj` and `TransformBuilder` implement the `Info` trait, which has read-only methods for retrieving data about the current `PROJ` setup; 
7. The distinction between the two structs is documented at the top level, and is hopefully clear?

Reply:
That all sounds reasonable to me! 

I'll close this PR and take another look at https://github.com/georust/proj/pull/34